R/crunch-data-frame.R

```diff
@@ -2,59 +2,66 @@
 #'
 #' CrunchDataFrames are designed to mimic the ways that [`data.frame`]s are used.
 #' They should be a drop-in replacement in many places where data.frames are used.
-#' 
-#' CrunchDataFrames are generated not by downloading all of the variables from 
-#' a dataset, but rather only the variables that are needed by subsequent 
-#' functions. So, if you create a CrunchDataFrame, and then run a linear model 
+#'
+#' CrunchDataFrames are generated not by downloading all of the variables from
+#' a dataset, but rather only the variables that are needed by subsequent
+#' functions. So, if you create a CrunchDataFrame, and then run a linear model
 #' using `lm()`, only the variables used by the linear model will be downloaded.
-#'  
-#' CrunchDataFrames can be altered (that is: adding more columns, removing 
-#' columns, subsetting rows, etc.) with the same `[`, `[[`, and `$` syntax as 
+#'
+#' CrunchDataFrames can be altered (that is: adding more columns, removing
+#' columns, subsetting rows, etc.) with the same `[`, `[[`, and `$` syntax as
 #' data.frames.
 #'
 #' @param x a CrunchDataFrame
 #' @param i indicators for rows (integers or logicals)
 #' @param j indicators for columns (names, integers, or logicals)
 #' @param value value to place (or replace) in the CrunchDataFrame
-#' @param drop logical. If `TRUE` the result is coerced to the lowest possible 
+#' @param drop logical. If `TRUE` the result is coerced to the lowest possible
 #' dimension. The default is to drop if only one column is left, but not to drop
 #' if only one row is left.
 #'
 #' @name CrunchDataFrame
 NULL
 
 CrunchDataFrame <- function (dataset, row.order = NULL,
-                             categorical.mode = "factor",
-                             include.hidden = FALSE) {
+    categorical.mode = "factor",
+    include.hidden = TRUE) {
     ## S3 constructor method for CrunchDataFrame. terms.formula doesn't seem
     ## to like S4 subclass of environment
     stopifnot(is.dataset(dataset))
-    
-    if (include.hidden) {
-        var_names <- aliases(allVariables(dataset))
-    } else {
-        var_names <- aliases(variables(dataset))
+
+    # When a user asks for a hidden variable with ds[, "hidden_var_name"] we should
+    # give it to them, and so overrride the include.hidden argument
+    allvars <- allVariables(dataset)
+    dsvars <- ShojiCatalog(crGET(self(allvars), query=list(relative="on")))
+    if (length(allvars) != length(dsvars) && # check that dataset has avariable subset
+            length(hiddenVariables(dataset)) > 0  # check that user is asking for hidden variable
+    ) {
+        include.hidden <- TRUE
     }
```

Comment:
I'm not sure you actually need this machinery. It looks like what you're trying to do is see if the dataset object that's being passed is a subset of the variables available, and if it is, and the dataset has any hidden variables then you set `include.hidden` to `TRUE`, but that shouldn't be necessary here: I don't think we need to override `include.hidden` here. We can just include them in the promises made later by using line 43 below, and then we can decide to fetch them or not in `as.data.frame.CrunchDataFrame`

Reply:
Ah, look at that. Thanks!